aoc2018/src/main/scala/day19.scala

```diff
@@ -0,0 +1,113 @@
+import scala.io.*
+import math.*
+import scala.collection.mutable
+import scala.collection.mutable.Stack
+import scala.util.{Failure, Success, Try}
+
+
+object day19 extends App:
+
+  private val day: String =
+    this.getClass.getName.drop(3).init
+
+  private val start1: Long =
+    System.currentTimeMillis
+
+  case class Operation(op: String, a: Int, b: Int, c: Int):
+    def addr(in: Array[Int]): Array[Int] = {in(c) = in(a) + in(b); in}
+    def addi(in: Array[Int]): Array[Int] = {in(c) = in(a) + b; in}
+    def mulr(in: Array[Int]): Array[Int] = {in(c) = in(a) * in(b); in}
+    def muli(in: Array[Int]): Array[Int] = {in(c) = in(a) * b; in}
+    def banr(in: Array[Int]): Array[Int] = {in(c) = in(a) & in(b); in}
+    def bani(in: Array[Int]): Array[Int] = {in(c) = in(a) & b; in}
+    def borr(in: Array[Int]): Array[Int] = {in(c) = in(a) | in(b); in}
+    def bori(in: Array[Int]): Array[Int] = {in(c) = in(a) | b; in}
+    def setr(in: Array[Int]): Array[Int] = {in(c) = in(a); in}
+    def seti(in: Array[Int]): Array[Int] = {in(c) = a; in}
+    def gtir(in: Array[Int]): Array[Int] = {in(c) = if a > in(b) then 1 else 0; in}
+    def gtri(in: Array[Int]): Array[Int] = {in(c) = if in(a) > b then 1 else 0; in}
+    def gtrr(in: Array[Int]): Array[Int] = {in(c) = if in(a) > in(b) then 1 else 0; in}
+    def eqir(in: Array[Int]): Array[Int] = {in(c) = if a == in(b) then 1 else 0; in}
+    def eqri(in: Array[Int]): Array[Int] = {in(c) = if in(a) == b then 1 else 0; in}
+    def eqrr(in: Array[Int]): Array[Int] = {in(c) = if in(a) == in(b) then 1 else 0; in}
+
+    val ops: Map[String, Array[Int] => Array[Int]] = Map(
+      "addr" -> addr, "addi" -> addi, "mulr" -> mulr, "muli" -> muli, "banr" -> banr, "bani" -> bani,
+      "borr" -> borr, "bori" -> bori, "setr" -> setr, "seti" -> seti, "gtir" -> gtir, "gtri" -> gtri,
+      "gtrr" -> gtrr, "eqir" -> eqir, "eqri" -> eqri, "eqrr" -> eqrr)
+
+    def run(input: Vector[Int]): Vector[Int] =
+      val executable: Array[Int] => Array[Int] = ops(op)
+      executable(input.toArray).toVector
+
+
+
+  private val input: (Vector[Operation], Int) =
+
+    def parseOps(s: String): Option[Operation] =
```

Comment:
parsing strings reads better with pattern matching, eg.:

```
s match
  case s"$op $i0 $i1 $i2" => Some(Operation(op, i0.toInt, i1.toInt, i2.toInt))
```

Reply:
Ah yes I cleaned it up using the pattern match on strings:)